### PR TITLE
fixing_filters

### LINKED ISSUE #93: Indeterminacy Locus- Filter needs fixing or removing to work with the new data
**Describe the bug**
After the latest changes to the database, the indeterminacy locus filter is breaking the connection to the database. This needs to be fixed. 

**To Reproduce**
1. Go to "Explore Systems"
2. Change the value in the "indeterminacy locus" filter
3. Apply Filters
4. Database connection breaks (Screenshot below)

image

**Expected behavior**
- If the function is irrelevant, you can remove it. Otherwise, fix the filter to work with the new data.
- The Database connection should not fail in such cases which needs the server to be restarted. Instead, the exception needs to be handled gracefully and the user should be able to continue to use the application fine. 
- The Error message should not mention anything about the docker. It can be something like, just an example. "Error connecting to the Database. Check with the system administrator"

### PR DESCRIPTION
**Pull request by lRolwes**
Fixes #93 

**What was changed?**
- Error Message Updated
- Indeterminacy Locus Dimension Filter Works
- If a server function fails now it will fail gracefully and you can just refresh the tab

**How it was Changed**

- Added Try/Except statements to allow PostgressConnector to fail gracefully
- Added a Reconnect function to PostgressConnector to re establish the server connection in case of a fail. Calles in getStatistics and getFilteredSystems
- Conditioned Indeterminacy Locus Dimension to work on current data until data is complete
- Updated Error message in ExploreSystems to no longer say Docker

**Why was it changed?**

- Indeterminacy Locus Dimension Filter causes the system to break and will not fetch results until the server is restarted
- Whenever the server breaks it must be restarted, the tab cannot just be refreshed

### CODE REVIEW

**Review by Evan-Richmond — APPROVED** [review]
Everything works as expected for me!

File: Backend/PostgresConnector.py [hunk 1]
```diff
@@ -17,87 +17,111 @@ def __init__(self):
                 password="dad_pass",
                 port = "5432")
 
+    def reconnect(self):
```
Comment (AbhilashKotha) [inline Backend/PostgresConnector.py:20]: We should not have this function. there is a better way to handle this.

File: Backend/PostgresConnector.py [hunk 2]
```diff
@@ -17,87 +17,111 @@ def __init__(self):
...
     # gets a system identified by its label, input is string
     def getSystem(self,label):
         columns = '*'
         sql = "SELECT " + columns + " FROM functions_dim_1_NF WHERE label = '" + label + "'"
         cur = self.connection.cursor()
-        cur.execute(sql)
-        result = cur.fetchall()
-        cur.close()
-        return result
+        try:
```
Comment (AbhilashKotha) [inline Backend/PostgresConnector.py:47]: try, except, and finally blocks are good. but in addition, we should use the "with" statement.

File: Backend/PostgresConnector.py [hunk 3]
```diff
@@ -17,87 +17,111 @@ def __init__(self):
...
+                                else:
+                                    poly += c[j][i] + mon[i]
+                                first_term = False
+                        if j == 0:
+                            poly += ' : '
+                    poly += ']'
+                    result.append([row[0], '1', row[1], poly, row[3]])
+                
+        except Exception as e:
+            self.reconnect()
```
Comment (AbhilashKotha) [inline Backend/PostgresConnector.py:118]: reconnect should not be the solution. We should not break the connection in the first place. Please look for the "rollback" command.

File: Frontend/src/pages/ExploreSystems.js [hunk 4]
```diff
@@ -160,7 +160,7 @@ function ExploreSystems() {
               }))
         } catch (error) {
             setSystems(null);
-            reportMajorError("Error: CANNOT CONNECT TO DATABASE: Make sure Docker is running correctly");
+            reportMajorError("Error: Issue fetching requested data");
```
Comment (AbhilashKotha) [inline Frontend/src/pages/ExploreSystems.js:163]: Can we make it more clear, something like..
```suggestion
            reportMajorError("There was an error while fetching the information requested. Please contact the system administrator.");
```

**Review by AbhilashKotha — CHANGES_REQUESTED** [review]
@lRolwes While the changes could work, I think there is a better way to handle this, could you please look at my suggestions. Feel free to reach out if you need more clarity. Thanks.

**Comment by AbhilashKotha** [comment]
@lRolwes Great, this is the approach we need! Just a few more improvements.

Wherever we have a cursor, we will use the "with" statement
Wherever we have a database query, we will use rollback in the except block

**Pull request merged**